controllers/machinedeployment_ondelete.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/controllers/mdutil"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// rolloutOnDelete implements the logic for the OnDelete MachineDeploymentStrategyType.
+func (r *MachineDeploymentReconciler) rolloutOnDelete(ctx context.Context, d *clusterv1.MachineDeployment, msList []*clusterv1.MachineSet) error {
+	newMS, oldMSs, err := r.getAllMachineSetsAndSyncRevision(ctx, d, msList, true)
+	if err != nil {
+		return err
+	}
+
+	// newMS can be nil in case there is already a MachineSet associated with this deployment,
+	// but there are only either changes in annotations or MinReadySeconds. Or in other words,
+	// this can be nil if there are changes, but no replacement of existing machines is needed.
+	if newMS == nil {
+		return nil
+	}
+
+	allMSs := append(oldMSs, newMS)
+
+	// Scale up, if we can.
+	if err := r.reconcileNewMachineSetOnDelete(ctx, allMSs, newMS, d); err != nil {
+		return err
+	}
+
+	if err := r.syncDeploymentStatus(allMSs, newMS, d); err != nil {
+		return err
+	}
+
+	// Scale down, if we can.
+	if err := r.reconcileOldMachineSetsOnDelete(ctx, oldMSs, allMSs, d); err != nil {
+		return err
+	}
+
+	if err := r.syncDeploymentStatus(allMSs, newMS, d); err != nil {
+		return err
+	}
+
+	if mdutil.DeploymentComplete(d, &d.Status) {
+		if err := r.cleanupDeployment(ctx, oldMSs, d); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+//reconcileOldMachineSetsOnDelete handles reconciliation of Old MachineSets associated with the MachineDeployment in the OnDelete MachineDeploymentStrategyType.
+func (r *MachineDeploymentReconciler) reconcileOldMachineSetsOnDelete(ctx context.Context, oldMSs []*clusterv1.MachineSet, allMSs []*clusterv1.MachineSet, deployment *clusterv1.MachineDeployment) error {
+	log := ctrl.LoggerFrom(ctx)
```

Comment:
Should we add a `strategy` key with `ondelete` value here, so it's added to all the log lines from here onward?